compiler/src/iree/compiler/Dialect/Util/Transforms/VerifyInitializationOrder.cpp

```diff
@@ -0,0 +1,407 @@
+                                 " accesses global '@" + globalName +
+                                 "' defined at position " +
+                                 Twine(globalOrdinal));
+              return WalkResult::interrupt();
+            }
+          }
+        }
+
+        // Also check stores for forward references.
+        if (auto storeOp = dyn_cast<IREE::Util::GlobalStoreOpInterface>(op)) {
+          auto globalName = storeOp.getGlobalName();
+          auto &global = globalTable.lookup(globalName);
+          if (global.op) {
+            size_t globalOrdinal = opOrdinalMap[global.op];
+            if (globalOrdinal > initOrdinal) {
+              emitDiagnostic(ERROR, op->getLoc(),
+                             "initializer at position " + Twine(initOrdinal) +
+                                 " stores to global '@" + globalName +
+                                 "' defined at position " +
+                                 Twine(globalOrdinal));
+              return WalkResult::interrupt();
+            }
+          }
+        }
+
+        // Check function calls for transitive dependencies.
+        if (auto callOp = dyn_cast<IREE::Util::CallOp>(op)) {
+          // We'll verify the called functions separately in a more
+          // sophisticated pass. For now, just ensure the function exists. DO
+          // NOT SUBMIT
+        }
+
+        return WalkResult::advance();
+      });
+      if (result.wasInterrupted()) {
+        return failure();
+      }
+    }
+    return success();
+  }
+
+  // Verifies that immutable globals are only initialized once.
+  LogicalResult verifyDoubleInitialization(
+      GlobalTable &globalTable,
+      ArrayRef<IREE::Util::InitializerOpInterface> initializerOps) {
+    // Track which immutable globals have been initialized.
+    llvm::StringMap<Location> immutableGlobalInitializations;
+
+    // First pass: check globals with initial values.
+    for (size_t i = 0; i < globalTable.size(); ++i) {
+      auto globalName = globalTable.lookupByOrdinal(i);
+      auto &global = globalTable.lookup(globalName);
+      if (global.op && !global.op.isGlobalMutable() &&
+          global.op.getGlobalInitialValue()) {
+        // Immutable global with initial value.
+        immutableGlobalInitializations.try_emplace(globalName,
+                                                   global.op->getLoc());
+      }
+    }
+
+    // Second pass: check for stores in initializers.
+    for (auto initializerOp : initializerOps) {
+      auto result = initializerOp.walk(
+          [&](IREE::Util::GlobalStoreOpInterface storeOp) -> WalkResult {
+            auto globalName = storeOp.getGlobalName();
+            auto &global = globalTable.lookup(globalName);
+            if (global.op && !global.op.isGlobalMutable()) {
+              // Check if already initialized.
+              auto it = immutableGlobalInitializations.find(globalName);
+              if (it != immutableGlobalInitializations.end()) {
```

Comment:
nit: maybe just `if (immutableGlobalInitializations.contains(globalName)) {`.


```suggestion
              if (immutableGlobalInitializations.contains(globalName)) {
```